internal/venice-common/src/main/java/com/linkedin/venice/utils/DiskUsage.java

```diff
@@ -85,8 +127,130 @@ public boolean isDiskFull(long bytesWritten) {
-        + " bytes free.";
+    StringBuilder sb = new StringBuilder();
+    if (databaseSizeLimiter != null) {
+      sb.append(databaseSizeLimiter.getDatabaseStatus()).append(", ");
+    }
+    sb.append("Disk at path: ")
+        .append(diskPath)
+        .append(" has ")
+        .append(totalSpaceBytes)
+        .append(" total bytes and ")
+        .append(freeSpaceBytes)
+        .append(" bytes free. ")
+        .append("Disk is marked 'full' when there are less than ")
+        .append(freeSpaceBytesRequired)
+        .append(" bytes free.");
+    return sb.toString();
+  }
+
+  @Override
+  public void close() throws IOException {
+    if (databaseSizeLimiter != null) {
+      databaseSizeLimiter.shutdown();
+    }
+  }
+
+  /**
+   * This class is used to measure whether the current database size exceeds the limit or not.
+   * It will measure the size after writing a configured amount of bytes, and the measurement
+   * won't be very accurate since the measurement is independent of the database write operation,
+   * and it was designed in this way to minimize the overhead of this class in the write path.
+   */
+  public static class DatabaseSizeLimiter {
+    private static final Logger LOGGER = LogManager.getLogger(DatabaseSizeLimiter.class);
+    public static final long DEFAULT_DATABASE_SIZE_MEASURE_INTERVAL = 256 * 1024 * 1024; // 256MB
+
+    private final String databasePath;
+    private final long databaseSizeLimit;
+    private final long databaseSizeMeasurementInterval;
+    private final AtomicLong currentDatabaseSize = new AtomicLong();
+    private final Supplier<Long> databaseSizeSupplier;
+    private final AtomicLong accumulatedByteSize = new AtomicLong();
+    private final AtomicBoolean hasOngoingMeasurement = new AtomicBoolean(false);
+    private final ExecutorService databaseSizeMeasurementExecutor =
+        Executors.newSingleThreadExecutor(new DefaultThreadFactory("Database_Size_Measurement"));
+    private boolean exceedDatabaseSizeLimit = false;
+
+    public DatabaseSizeLimiter(String databasePath, long databaseSizeLimit, long databaseSizeMeasurementInterval) {
+      this(
+          databasePath,
+          databaseSizeLimit,
+          databaseSizeMeasurementInterval,
+          () -> FileUtils.sizeOf(new File(databasePath)));
+    }
+
+    // For testing
+    public DatabaseSizeLimiter(
+        String databasePath,
+        long databaseSizeLimit,
+        long databaseSizeMeasurementInterval,
+        Supplier<Long> databaseSizeSupplier) {
+      this.databasePath = databasePath;
+      this.databaseSizeLimit = databaseSizeLimit;
+      this.databaseSizeMeasurementInterval = databaseSizeMeasurementInterval;
+      this.databaseSizeSupplier = databaseSizeSupplier;
+    }
+
+    public boolean hitDatabaseSizeLimit(long bytesWritten) {
+      if (databaseSizeLimit <= 0) {
+        return false;
+      }
+      if (exceedDatabaseSizeLimit) {
```

Comment:
Looks like once it is set to true, there is no way to set it back to false? Then hybrid store ingestion will be paused but not resumed?

Reply:
At runtime, no.
Maybe this is too brutal, I guess we can keep the ingestion retrying at least for hybrid stores since a failed push could delete the new version, so that the existing version can resume.

Reply:
cc @anaberezhnov

Reply:
My understanding that the new check will run both in parent and in ingestion isolation processes. So both can get stuck as a result.

Reply:
Let's also add an alertable metric so that user can take action immediately.

Reply:
If we are going to add this safeguard, let's also introduce the jitter into the limiter so not every host will get into the trouble at the same time. Along with the alertable metric, customers can get alerted on these nodes with lower limit first and take actions.